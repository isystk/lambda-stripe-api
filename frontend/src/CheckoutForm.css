.checkout-form {
  display: flex;
  flex-direction: column;
  margin-bottom: 1rem;
}

label {
  font-size: 1rem;
  margin-bottom: 0.5rem;
}

input, .buy-btn {
  font-size: 1rem;
  padding: 0.5rem;
  border-radius: 4px;
  border: 1px solid #ccc;
  box-shadow: none;
}

.card-element-wrapper {
  padding: 16px;
  background-color: #eee;
  border-radius: 4px;
}

.error {
  color: red;
}