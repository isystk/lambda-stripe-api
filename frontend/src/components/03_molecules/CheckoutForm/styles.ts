import { css, keyframes } from '@emotion/css'

export const checkoutForm = css``
