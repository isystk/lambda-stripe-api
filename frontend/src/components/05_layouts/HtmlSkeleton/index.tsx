import React, { Children, FC, useState } from 'react'
import { connect } from '@/components/hoc'
import { ContainerProps, WithChildren } from 'types'
import useAppRoot from '@/stores/useAppRoot'
import Title, { TitleProps } from './Title'
import Head from 'next/head'
import { isReactElement } from '@/utils/general/object'
import { APP_NAME, APP_DESCRIPTION } from '@/constants'

/** HtmlSkeleton Props */
export type HtmlSkeletonProps = WithChildren & {
  showSideMenu?: boolean
}
/** Presenter Props */
export type PresenterProps = HtmlSkeletonProps & {
  title?: TitleProps['children']
  description: string
}

/** Presenter Component */
const HtmlSkeletonPresenter: FC<PresenterProps> = ({
  showSideMenu,
  main,
  title,
  description,
  children,
}) => (
  <>
    <Head>
      {/* タイトル */}
      <title>{title}</title>
      {/* favicon */}
      <link rel="icon" href="/favicon.ico" />
      {/* PCやモバイル（スマホ、タブレット）などのデバイスごとのコンテンツの表示領域 */}
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      {/* ブラウザテーマカラー */}
      <meta name="theme-color" content="#ffffff" />
      {/* サイト概要 */}
      <meta name="description" content={description} />
      {/* OGP 画像URL */}
      <meta property="og:image" content={'/ogp-image.png'} />
      {/* OGP タイトル */}
      <meta name="og:title" content={title} />
      {/* OGP サイト概要 */}
      <meta name="og:description" content={description} />
      {/* OGP Twitterカード */}
      <meta name="twitter:card" content="summary" />
      {/* apple ポータブル端末 アイコン */}
      <link rel="apple-touch-icon" href="/apple-touch-icon.png" />
      {/* manifest.json */}
      <link rel="manifest" href="/manifest.json" />
      {/* Tailwind UI で採用されているカスタムフォント */}
      <link rel="stylesheet" href="https://rsms.me/inter/inter.css" />
    </Head>
    <Context.Provider value={main}>{children}</Context.Provider>
  </>
)

/** Container Component */
const HtmlSkeletonContainer: React.FC<
  ContainerProps<HtmlSkeletonProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  let title: TitleProps['children'] | undefined = undefined
  const main = useAppRoot()

  if (!main) return <></>

  children = Children.map(children, (child) =>
    isReactElement(child) && child.type === Title
      ? (title = `${child.props.children} | ${APP_NAME}`) && undefined
      : child
  )
  if (!title) title = APP_NAME

  const description = APP_DESCRIPTION
  return presenter({
    main,
    title,
    description,
    children,
    ...props,
  })
}

export const Context = React.createContext(null)

export default connect<HtmlSkeletonProps, PresenterProps>(
  'HtmlSkeleton',
  HtmlSkeletonPresenter,
  HtmlSkeletonContainer
)

// Sub Component
export type { TitleProps }
export { Title }
