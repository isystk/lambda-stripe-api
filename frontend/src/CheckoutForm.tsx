import { CardElement, useElements, useStripe } from "@stripe/react-stripe-js";
import React, { useState } from "react";
import "./CheckoutForm.css"

const CARD_ELEMENT_OPTIONS = {
  hidePostalCode: true, // 郵便番号を非表示
  style: {
      base: {
        fontFamily: 'Sans Serif',
        fontStyle: 'normal',
        fontWeight: '400',
        fontSize: '16px',
        letterSpacing: '1px',
        color: '#333333',
        '::placeholder': {
          color: '#a9a9a9'
        }
      }
  }
};
function CheckoutForm() {
  
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const productId = process.env.REACT_APP_PRODUCT_ID;
  
  const stripe = useStripe();
  const elements = useElements();

  const createSubscription = async () => {
    try {
      
      // create a payment method
      const paymentMethod = await stripe?.createPaymentMethod({
        type: "card",
        card: elements?.getElement(CardElement)!,
        billing_details: {
          name,
          email,
        },
      });

      // call the backend to create subscription
      const response = await fetch(process.env.REACT_APP_PAYMENT_API_URL??'', {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          paymentMethod: paymentMethod?.paymentMethod?.id,
          name,
          email,
          productId
        }),
      }).then((res) => res.json());

      const confirmPayment = await stripe?.confirmCardPayment(
        response.clientSecret
      );

      if (confirmPayment?.error) {
        alert(confirmPayment.error.message);
      } else {
        alert("Success! Check your email for the invoice.");
      }
    } catch (error) {
      console.log(error);
    }
  };
  
  return (
    <div>
        <div>
            <label htmlFor="name" >名前:</label>
            <input
                id="name"
                type="text"
                value={name}
                onChange={(e) => setName(e.target.value)}
            />
        </div>
        <div>
            <label htmlFor="email">メールアドレス:</label>
            <input
                id="email"
                type="text"
                value={email}
                onChange={(e) => setEmail(e.target.value)}
            />
        </div>
        <div className="card-element-wrapper">
            <CardElement options={CARD_ELEMENT_OPTIONS} />
        </div>
        <div>
            <button className="buy-btn" onClick={createSubscription} disabled={!stripe}>購入する</button>
        </div>
    </div>
  );
}

export default CheckoutForm;