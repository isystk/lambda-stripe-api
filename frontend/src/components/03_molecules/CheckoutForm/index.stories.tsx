import { Meta, Story } from '@storybook/react'
import React from 'react'
import CheckoutForm, { CheckoutFormProps } from './index'

export default {
  title: '03_molecules/CheckoutForm',
  component: CheckoutForm,
} as Meta

const Template: Story = () => {
  // const props: CheckoutFormProps = {
  //   product: {
  //     plans: [
  //       {
  //         id: 1, amount: 999, currency: 'jpy'
  //       }
  //     ]
  //   }
  // }
  // return (
  //     <CheckoutForm {...props} />
  // )
  return <></>
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
