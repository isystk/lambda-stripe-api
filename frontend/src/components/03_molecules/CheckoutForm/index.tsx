import React, { FC, useContext, useState } from 'react'
import { ContainerProps, WithChildren } from 'types'
import * as styles from './styles'
import { connect } from '@/components/hoc'
import MainService from '@/services/main'
import { Context } from '@/components/05_layouts/HtmlSkeleton'
import Loading from '@/components/01_atoms/Loading'
import { CardElement, useElements, useStripe } from '@stripe/react-stripe-js'
import { useForm, type SubmitHandler } from 'react-hook-form'
import axios, { AxiosError } from 'axios'
import { PaymentMethodResult } from '@stripe/stripe-js'
import { Api } from '@/constants/api'
import * as stripe from 'stripe'

const CARD_ELEMENT_OPTIONS = {
  hidePostalCode: true, // 郵便番号を非表示
  style: {
    base: {
      fontFamily: 'Sans Serif',
      fontStyle: 'normal',
      fontWeight: '400',
      fontSize: '16px',
      letterSpacing: '1px',
      color: '#333333',
      '::placeholder': {
        color: '#a9a9a9',
      },
    },
  },
}

type FormInputs = {
  name: string
  email: string
  planId: string
}

export type ProductData = {
  plans: stripe.Plan[]
} & stripe.Product

/** CheckoutFormProps Props */
export type CheckoutFormProps = WithChildren & {
  product: ProductData
  userKey?: string
}
/** Presenter Props */
export type PresenterProps = CheckoutFormProps & {
  main
  onsubmit
  product
  isComplete
  loading
  cardErrorMsg
  errorMsg
  register
  handleSubmit
  errors
  stripe
}

/** Presenter Component */
const CheckoutFormPresenter: FC<PresenterProps> = ({
  main,
  onsubmit,
  product,
  isComplete,
  loading,
  cardErrorMsg,
  errorMsg,
  register,
  handleSubmit,
  errors,
  validate,
  stripe,
  ...props
}) => (
  <>
    <div>
      {!isComplete ? (
        <form onSubmit={handleSubmit(onsubmit)}>
          <div className="flex flex-wrap items-center md:mb-8">
            {product.plans.map(({ id, amount, currency }, idx) => {
              const amountFmt = amount
                ? new Intl.NumberFormat('ja-JP', {
                    style: 'currency',
                    currency,
                  }).format(amount)
                : ''
              return (
                <div key={id} className="w-full md:w-1/2">
                  <div className="bg-black rounded-lg md:mr-4 py-12 h-72 mb-8">
                    <p className="text-white text-center mb-12">プラン</p>
                    <p className="text-white font-bold text-center text-5xl mb-12">
                      {amountFmt}
                    </p>
                    <input
                      type="radio"
                      checked={idx === 0}
                      value={id}
                      {...register('planId', validate['planId'])}
                    />
                  </div>
                </div>
              )
            })}
            {errors.planId && (
              <span className="pt-4 text-red-500">{errors.planId.message}</span>
            )}
          </div>
          <div className="flex flex-col mb-4">
            <input
              placeholder="お名前"
              type="text"
              className="p-3 bg-gray-200 rounded-md"
              {...register('name', validate['name'])}
            />
            {errors.name && (
              <span className="pt-4 text-red-500">{errors.name.message}</span>
            )}
          </div>
          <div className="flex flex-col mb-4">
            <input
              placeholder="メールアドレス"
              type="email"
              className="p-3 bg-gray-200 rounded-md"
              {...register('email', validate['email'])}
            />
            {errors.email && (
              <span className="pt-4 text-red-500">{errors.email.message}</span>
            )}
          </div>
          <div className="p-4 bg-gray-200 rounded-md">
            <CardElement options={CARD_ELEMENT_OPTIONS} />
            {cardErrorMsg && (
              <span className="pt-4 text-red-500">{cardErrorMsg}</span>
            )}
          </div>
          <div>
            <button
              className="bg-blue-500 rounded-md text-white border-none rounded-5 px-20 py-4 text-18 cursor-pointer mt-10"
              type="submit"
              disabled={!stripe}
            >
              購入する
            </button>
          </div>
          {errorMsg && <p className="pt-4 text-red-500">{errorMsg}</p>}
        </form>
      ) : (
        <p className="text-18 text-left mb-20">購入が完了しました。</p>
      )}
      <Loading loading={loading} />
    </div>
  </>
)

/** Container Component */
const CheckoutFormContainer: React.FC<
  ContainerProps<CheckoutFormProps, PresenterProps>
> = ({ presenter, children, product, userKey, ...props }) => {
  const main = useContext<MainService | null>(Context)
  if (!main) return <></>

  const [isComplete, setIsComplete] = useState(false)
  const [loading, setLoading] = useState(false)
  const [cardErrorMsg, setCardErrorMsg] = useState('')
  const [errorMsg, setErrorMsg] = useState('')
  const stripe = useStripe()
  const elements = useElements()

  const {
    register,
    handleSubmit,
    reset,
    formState: { errors },
  } = useForm<FormInputs>()

  const validate = {
    planId: {
      required: 'プランを選択してください',
    },
    name: {
      required: 'お名前を入力してください',
      max: { value: 30, message: 'お名前は３０文字以内で入力してください' },
    },
    email: {
      required: 'メールアドレスを入力してください',
      pattern: {
        value: /[\w\-\\._]+@[\w\-\\._]+\.[A-Za-z]+/,
        message: 'メールアドレスを正しく入力してください',
      },
    },
  }

  // フォーム送信ボタンを押された時の処理
  const onsubmit: SubmitHandler<FormInputs> = async ({
    planId,
    name,
    email,
  }) => {
    try {
      // create a payment method
      const payment: PaymentMethodResult | undefined =
        await stripe?.createPaymentMethod({
          type: 'card',
          card: elements?.getElement(CardElement),
          billing_details: {
            email,
          },
        })
      if (!payment) {
        setCardErrorMsg('カード情報を入力してください')
        return
      }
      if (payment.error) {
        setCardErrorMsg(payment.error.message ?? 'カード情報が不正です')
        return
      }
      setLoading(true)

      // 決済処理をする
      const { data } = await axios.post(Api.Payment, {
        userKey: userKey ?? email,
        paymentMethod: payment.paymentMethod?.id,
        name,
        email,
        planId,
      })
      reset()

      setIsComplete(true)
    } catch (e: unknown) {
      console.log(e)
      if (e instanceof AxiosError) {
        const { response } = e
        setErrorMsg(response?.data?.message)
      } else if (e instanceof Error) {
        setErrorMsg(e.message)
      }
    } finally {
      setLoading(false)
    }
  }

  return presenter({
    children,
    main,
    onsubmit,
    product,
    isComplete,
    loading,
    cardErrorMsg,
    errorMsg,
    register,
    handleSubmit,
    errors,
    validate,
    stripe,
    ...props,
  })
}

export default connect<CheckoutFormProps, PresenterProps>(
  'CheckoutForm',
  CheckoutFormPresenter,
  CheckoutFormContainer
)
