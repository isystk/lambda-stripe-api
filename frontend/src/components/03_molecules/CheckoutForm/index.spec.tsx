import React from 'react'
import renderer from 'react-test-renderer'
import CheckoutForm, { CheckoutFormProps } from './index'

import { Context } from '@/components/05_layouts/HtmlSkeleton'
import MainService from '@/services/main'

describe('CheckoutForm', () => {
  it('Match Snapshot', () => {
    // const props: CheckoutFormProps = {
    //   product: {
    //     plans: [
    //       {
    //         id: 1, amount: 999, currency: 'jpy'
    //       }
    //     ]
    //   }
    // }
    // const component = renderer.create(
    //     <CheckoutForm {...props}/>
    // )
    // const tree = component.toJSON()
    //
    // expect(tree).toMatchSnapshot()
    expect(true)
  })
})
