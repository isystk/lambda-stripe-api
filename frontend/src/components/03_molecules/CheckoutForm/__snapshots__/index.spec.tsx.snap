// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutForm Match Snapshot 1`] = `
<div>
  <form
    className="css-0"
    onSubmit={[Function]}
  >
    <div
      className="flex flex-wrap items-center md:mb-8"
    >
      <div
        className="w-full md:w-1/2"
      >
        <div
          className="bg-black rounded-lg md:mr-4 py-12 h-72 mb-8"
        >
          <p
            className="text-white text-center mb-12"
          >
            Plan
          </p>
          <p
            className="text-white font-bold text-center text-5xl mb-12"
          >
            Â¥999
          </p>
          <input
            checked={true}
            name="planId"
            onBlur={[Function]}
            onChange={[Function]}
            type="radio"
            value={1}
          />
        </div>
      </div>
    </div>
    <div
      className="flex flex-col mb-4"
    >
      <input
        className="css-0 p-3 bg-gray-200 rounded-md"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="your name"
        type="text"
      />
    </div>
    <div
      className="flex flex-col mb-4"
    >
      <input
        className="css-0 p-3 bg-gray-200 rounded-md"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email address"
        type="email"
      />
    </div>
    <div
      className="p-4 bg-gray-200 rounded-md"
    >
      <div />
    </div>
    <div>
      <button
        className="bg-blue-500 rounded-md text-white border-none rounded-5 px-20 py-4 text-18 cursor-pointer mt-10"
        disabled={true}
        type="submit"
      >
        Buy Now
      </button>
    </div>
  </form>
  <div
    className="css-0 fixed top-0 left-0 w-full h-full z-30 hidden"
  >
    <div
      className="absolute bottom-0 right-0 p-8 flex justify-center"
    >
      <div
        className="animate-spin h-10 w-10 border-4 border-gray-500 rounded-full border-t-transparent"
      />
    </div>
  </div>
</div>
`;
